keps/sig-api-machinery/20200421-static-admission-webhook-configuration.md

```diff
@@ -0,0 +1,149 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Static Webhook configuration allows registering admission webhooks during
+kube-apiserver start up allowing for no delays in policy enforcement between
+policy addition and kube-apiserver startup. 
+
+## Motivation
+
+Today most policy enforcement is implemented through MutatingAdmissionWebhooks
+and/or ValidatingAdmissionWebhooks. These admission webhooks are registered
+through creating MutatingWebhookConfiguration or ValidatingWebhookConfiguration
+objects. Any policy enforcement is not in place until these webhook
+configurations are created, thereby registering the webhook. This creates a gap
+in enforcement spanning from when the kube-apiserver is started to until the
+webhook configuration objects are created and picked up by the dynamic admission
+controller. Another gap is the inability of the cluster administrator to protect
+against deletion of these webhook configuration objects as
+MutatingWebhookConfiguration and ValidatingWebhookConfiguration objects are not
+subject to webhook admission policy.
+
+This KEP aims to address these issues.  
+
+### Goals
+
+- A robust admission webhook registration process where there is no period of
+  time between the kube-apiserver coming up and the registration of an admission
+  webhook.
+
+- The registration of webhooks registered through this process should be
+  protected from alteration by API requests made by cluster users.
+
+- The new registration process should also have a highly privileged means of
+  being able to alter webhook registrations.
+ 
+## Proposal
+
+In short, this proposal is about augmenting `AdmissionConfig`'s plugin
+`configuration` [specification](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers)
+to include a path to a configuration file containing one of the following:
+- `admissionregistration.k8s.io/v1.ValidatingWebhookConfigurationList` 
+- `admissionregistration.k8s.io/v1.MutatingWebhookConfigurationList` 
+- v1.List with items of admissionregistration.k8s.io/v1.ValidatingWebhookConfiguration 
+- v1.List  with items of admissionregistration.k8s.io/v1.MutatingWebhookConfiguration.
+This configuration is loaded  as static webhooks in api servers. These
+webhooks are called by validating and mutating admission plugins along with
+dynamically loaded webhooks for relevant admission requests. This also means
+that these webhooks are not API visible objects and hence cannot be modified by
+an API user.
```

Comment:
Are the static webhooks ordered differently from the dynamic webhooks? Or just mixed in? This is mostly only relevant to mutating hooks.

Reply:
Manifest based webhooks won't provide any guarantees about ordering. I feel that it would be out of the scope of the KEP.

Please let me know if you feel differently